### ISSUE TITLE
Simulation of large simulation crashes with I/O errors

### ISSUE THREAD

**Issue by mschmidt87** [jakobj/nest-simulator#82]
I simulated the multi-area model [1] (see here for the code, contact me if you need access rights to the repository) with this branch and ended up with I/O errors. I used commit e760d1d9611da99aa0b459c9122ce428f0b02631 .

Setup: Simulation was run on the local HPC cluster (Blaustein, Juelich) on 720 MPI processes (no threading) with 1 spike detector for the entire network, i.e. 720 spike files were created.
The simulation crashed with the output as posted below. I ran another test simulation with a down-scaled network (1% of neurons and 1% of indegree per neuron) on 4 MPI processes (no threading) and this simulation did not produce any errors.

Here is an excerpt of the error file:

> Sorry!  You were supposed to get help about:
>     mpi_errors_are_fatal
> But I couldn't open the help file:
>     /opt/software/pystuff_new/share/openmpi/help-mpi-errors.txt: Too many open files.  Sorry!
> --------------------------------------------------------------------------
> [blaustein17][[47852,1],72][btl_tcp_component.c:1102:mca_btl_tcp_component_accept_handler] --------------------------------------------------------------------------
> Sorry!  You were supposed to get help about:
>     mpi_errors_are_fatal
> But I couldn't open the help file:
>     /opt/software/pystuff_new/share/openmpi/help-mpi-errors.txt: Too many open files.  Sorry!
> --------------------------------------------------------------------------
> [...]
> 
> 
> [blaustein21][[47852,1],261][btl_tcp_component.c:1102:mca_btl_tcp_component_accept_handler] accept() failed: Too many open files (24).[blaustein16][[47852,1],0][btl_tcp_component.c:1102:mca_btl_tcp_component_accept_handler] accept() failed: Too many open files (24).
> [blaustein16:23344] *** Process received signal ***
> [blaustein16:23344] Signal: Segmentation fault (11)
> [blaustein16:23344] Signal code: Address not mapped (1)
> [blaustein16:23344] Failing at address: (nil)
> [blaustein16:23344] [ 0] /lib64/libpthread.so.0(+0xf130)[0x7ff167037130]
> [blaustein16:23344] *** End of error message ***
> [blaustein16][[47852,1],1][btl_tcp_component.c:1102:mca_btl_tcp_component_accept_handler] accept() failed: Too many open files (24).
> [blaustein16:23345] *** Process received signal ***
> [blaustein16:23345] Signal: Segmentation fault (11)
> [blaustein16:23345] Signal code: Address not mapped (1)
> [blaustein16:23345] Failing at address: (nil)
> [blaustein16:23345] [ 0] /lib64/libpthread.so.0(+0xf130)[0x7fcf9583b130]
> [blaustein16:23345] *** End of error message ***
>

**Comment by jakobj** [comment]
@jougs @apeyser have you encountered this before, maybe during NEST IO development?

**Comment by heplesser** [comment]
@mschmidt87 I would think that writing 720 files in parallel would kill most file systems and I would assume that Blaustein has not really been optimized for handling such massive output. Does the network run on 720 processes without any output to file?

**Comment by mschmidt87** [comment]
@heplesser , okay, I did not expect that. I have successfully run jobs with 254 jobs (1 spike detector per population in the network, using global spike detector mode in the current master branch) on the same system, thus I did not expect to face principal problems with 720 files.

So if that is true, my conclusion is that the 5g branch without the NEST IO development merged will not be usable for the simulations of large networks.

I will run the proposed test and get back with the results.

**Comment by apeyser** [comment]
This sounds like something that @lekshmideepu and sandra.  I think turning off the SLI frontend was the fix.

**Comment by heplesser** [comment]
@mschmidt87 Take a look at https://stackoverflow.com/questions/880557/socket-accept-too-many-open-files, that may have useful information. It may be that the system has a limit on the number of simultaneously open files, and if that is 256 or 512, your 254-proc job may have been fine, while the 720-proc job fails. I assume MAM opens a number of files (various SLI files to read), and if 720 processes do that in parallel, this may cause problems. Indeed, during startup, each of the 720 processes has to read a lot of *.sli files, which may lead to trouble in itself.

Can you try with a smaller number of processes and rather use more threads, or on a different system? Or change the file limits as in the StackOverflow post?

**Comment by mschmidt87** [comment]
I ran the commands from the link and obtained these results:
`
cat /proc/sys/fs/file-max`
--> 13094743

`ulimit -n`
-> 1024

It's true that even without any spike files, the simulation is opening a lot of python files (I am now using PyNEST, no more SLI), but this is also true for the simulation with the NEST master, so it doesn't make sense to me that this system settings causes a problem for 5G only?

I ran a simulation with 24 MPI processes with 30 threads each (no output spike files) and this does not crash.

**Comment by heplesser** [comment]
@jakobj @suku248 Any idea why Max may see these problems only with 5G?

**Comment by jakobj** [comment]
from the top of my head, no. we haven't touched IO, so i'm not sure what could cause this problem. need to think about it a bit more.

**Comment by jakobj** [comment]
@jarsi will report once results from large scale runs are available

**Comment by jakobj** [comment]
@jarsi any news on this?

**Comment by jarsi** [comment]
I haven't had time yet to look into it but I will do so as soon as possible. I have tested the mesocircuit with 5g on Jureca and this worked well.

**Comment by heplesser** [comment]
We (@jarsi) has now run a number of successful benchmarks on Jureca with the MAM model. The main problem we encountered were out-of-memory errors, which we have addressed by different memory allocation strategies. I will shortly post a PR. Since those simulations (when not affected by out-of-memory) were stable, for 1 and 4 threads per process, I consider this issue solved, once the new allocation strategies are merged.
